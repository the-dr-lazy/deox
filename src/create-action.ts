export type Action<
  TType extends string,
  TPayload = undefined,
  TMeta = undefined
> = TPayload extends undefined
  ? (TMeta extends undefined ? { type: TType } : { type: TType; meta: TMeta })
  : (TPayload extends Error
      ? (TMeta extends undefined
          ? { type: TType; payload: TPayload; error: true }
          : { type: TType; payload: TPayload; meta: TMeta; error: true })
      : (TMeta extends undefined
          ? { type: TType; payload: TPayload }
          : { type: TType; payload: TPayload; meta: TMeta }))
export type AnyAction = Action<string>

/**
 * Minimal (type-only) action factory
 * @example
 * const clearTodos = action('[Todo] truncate');
 */
export function createAction<TType extends string>(type: TType): Action<TType>
/**
 * Action with error factory
 * @example
 * const fetchTodosRejected = (payload: Error) => action('[Todo] fetch rejected', payload);
 */
export function createAction<TType extends string, TPayload extends Error>(
  type: TType,
  payload: TPayload
): Action<TType, TPayload>
/**
 * Action with non-error payload factory
 * @example
 * const addTodo = ({ name, completed = false }: Todo) => action('[Todo] add', { name, completed });
 */
export function createAction<TType extends string, TPayload>(
  type: TType,
  payload: TPayload
): Action<TType, TPayload>
/**
 * Action with error and meta factory
 * @example
 * const fetchTodosRejected = (payload: Error, meta?: Meta) => action('[Todo] fetch rejected', payload, meta);
 */
export function createAction<
  TType extends string,
  TPayload extends Error,
  TMeta
>(type: TType, payload: TPayload, meta: TMeta): Action<TType, TPayload, TMeta>
/**
 * Action with payload and meta factory
 * @example
 * const addTodo = ({ name, completed = false }: Todo, meta?: Meta) => action('[Todo] add', { name, completed }, meta);
 */
export function createAction<TType extends string, TPayload, TMeta>(
  type: TType,
  payload: TPayload,
  meta: TMeta
): Action<TType, TPayload, TMeta>

/**
 * Flux standard action factory
 * @example
 * const clearTodos = action('[Todo] truncate');
 * @example
 * const fetchTodosRejected = (payload: Error) => action('[Todo] fetch rejected', payload);
 * @example
 * const addTodo = ({ name, completed = false }: Todo) => action('[Todo] add', { name, completed });
 * @example
 * const fetchTodosRejected = (payload: Error, meta?: Meta) => action('[Todo] fetch rejected', payload, meta);
 * @example
 * const addTodo = ({ name, completed = false }: Todo, meta?: Meta) => action('[Todo] add', { name, completed }, meta);
 */
export function createAction<TType extends string, TPayload, TMeta>(
  type: TType,
  payload?: TPayload,
  meta?: TMeta
) {
  return {
    type,
    ...(payload !== undefined ? { payload } : {}),
    ...(meta !== undefined ? { meta } : {}),
    ...(payload instanceof Error ? { error: true } : {}),
  }
}
