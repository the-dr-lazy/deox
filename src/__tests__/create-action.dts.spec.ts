import { createAction } from '../create-action'

// @dts-jest:group action

// @dts-jest:pass:snap
createAction('[Todo] truncate')

// @dts-jest:pass:snap
createAction('[Todo] fetch rejected', new Error())

// @dts-jest:pass:snap
createAction('[Todo] fetch rejected', new Error(), { status: 403 })

// @dts-jest:pass:snap
createAction('[Todo] add', { name: `Read Robbin's` })

// @dts-jest:pass:snap
createAction('[Todo] add', { name: `Read Robbin's` }, 'Book is on bookshelf')

// @dts-jest:pass:snap
createAction('[Todo] add', { name: `Read Robbin's` }, 'Book is on bookshelf')
