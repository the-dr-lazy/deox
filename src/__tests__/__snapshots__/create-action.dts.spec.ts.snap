// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action createAction('[Todo] add', { name: \`Read Robbin's\` }) (type) should match snapshot 1`] = `"{ type: \\"[Todo] add\\"; payload: { name: string; }; }"`;

exports[`action createAction('[Todo] add', { name: \`Read Robbin's\` }, 'Book is on bookshelf') (type) should match snapshot 1`] = `"{ type: \\"[Todo] add\\"; payload: { name: string; }; meta: string; }"`;

exports[`action createAction('[Todo] add', { name: \`Read Robbin's\` }, 'Book is on bookshelf') (type) should match snapshot 2`] = `"{ type: \\"[Todo] add\\"; payload: { name: string; }; meta: string; }"`;

exports[`action createAction('[Todo] fetch rejected', new Error()) (type) should match snapshot 1`] = `"{ type: \\"[Todo] fetch rejected\\"; payload: Error; error: true; }"`;

exports[`action createAction('[Todo] fetch rejected', new Error(), { status: 403 }) (type) should match snapshot 1`] = `"{ type: \\"[Todo] fetch rejected\\"; payload: Error; meta: { status: number; }; error: true; }"`;

exports[`action createAction('[Todo] truncate') (type) should match snapshot 1`] = `"{ type: \\"[Todo] truncate\\"; }"`;
